WebRoot/js/zimbraAdmin/servers/view/ZaNewVolumeXWizard.js

```diff
@@ -0,0 +1,1062 @@
+
+    this._localXForm.setController();
+    this._localXForm.addListener(
+        DwtEvent.XFORMS_FORM_DIRTY_CHANGE,
+        new AjxListener(this, ZaNewVolumeXWizard.prototype.handleXFormChange)
+    );
+    this._localXForm.addListener(
+        DwtEvent.XFORMS_VALUE_ERROR,
+        new AjxListener(this, ZaNewVolumeXWizard.prototype.handleXFormChange)
+    );
+    this._helpURL =
+        location.pathname +
+        ZaUtil.HELP_URL +
+        "managing_servers/adding_a_new_storage_volume_to_the_server.htm?locid=" +
+        AjxEnv.DEFAULT_LOCALE;
+};
+
+ZaNewVolumeXWizard.bucketChoices = new XFormChoices([], XFormChoices.OBJECT_LIST, "globalBucketUUID", "bucketName");
+ZaNewVolumeXWizard.stepObject = new Object();
+ZaNewVolumeXWizard.prototype = new ZaXWizardDialog();
+ZaNewVolumeXWizard.prototype.constructor = ZaNewVolumeXWizard;
+ZaNewVolumeXWizard.prototype.toString = function () {
+    return "ZaNewVolumeXWizard";
+};
+ZaXDialog.XFormModifiers["ZaNewVolumeXWizard"] = new Array();
+ZaNewVolumeXWizard.helpURL =
+    location.pathname + ZaUtil.HELP_URL + "managing_accounts/create_an_account.htm?locid=" + AjxEnv.DEFAULT_LOCALE;
+ZaNewVolumeXWizard.prototype.handleXFormChange = function () {
+    if (this._localXForm.hasErrors()) {
+        this._button[DwtWizardDialog.FINISH_BUTTON].setEnabled(false);
+    } else {
+        // Validate new volume step fields
+        if (this._containedObject[ZaModel.currentStep] !== ZaNewVolumeXWizard.GENERAL_STEP) {
+            var areCommonVolumeFieldsEmpty = this.hasEmptyFields([ZaServer.A_VolumeName, ZaServer.A_VolumeType]);
+            var areInternalVolumeFieldsEmpty = this.hasEmptyFields([
+                ZaServer.A_VolumeRootPath,
+                ZaServer.A_VolumeCompressionThreshold,
+            ]);
+            var areExternalVolumeFieldsEmpty = this.hasEmptyFields([
+                ZaServer.A_VolumePrefix,
+                ZaServer.A_CompatibleS3Bucket,
+            ]);
+
+            if (
+                areCommonVolumeFieldsEmpty ||
+                (this._containedObject[ZaModel.currentStep] === ZaNewVolumeXWizard.NEW_INTERNAL_VOLUME &&
+                    areInternalVolumeFieldsEmpty) ||
+                (this._containedObject[ZaModel.currentStep] !== ZaNewVolumeXWizard.NEW_INTERNAL_VOLUME &&
+                    areExternalVolumeFieldsEmpty)
+            ) {
+                this._button[DwtWizardDialog.FINISH_BUTTON].setEnabled(false);
+            } else {
+                this._button[DwtWizardDialog.FINISH_BUTTON].setEnabled(true);
+            }
+        }
+
+        // Validate new bucket step fields
+        if (this._containedObject[ZaModel.currentStep] === ZaNewVolumeXWizard.NEW_S3_BUCKET) {
+            this._button[DwtWizardDialog.NEXT_BUTTON].setEnabled(false);
+        }
+    }
+};
+
+ZaNewVolumeXWizard.prototype.hasEmptyFields = function (fieldArray) {
+    var result = false;
+    for (var i in fieldArray) {
```

Comment:
for normal array traversal, it's better to use for(i = 0 ... ) form, so we don't have to care about prototype properties

Reply:
done